File: dp/fib.cpp
/* 
 * Top-down dynamic memoization implementation 
 * of the fibonacci sequence. 
 *
 * ASIDE: (As somebody who's done research on the fib sequence...)
 * Who will ever prove the largeness of the fibonacci sequence?
 * I tried several times, to no avail.
 * I thought when I obtained all this knowledge, 
 * that I would be able to compute a meaningful answer,
 * but I only ended up realizing that computation was just a tool,
 * not a way of solving problems.
 *
 * If somebody can ever prove the largeness of the fibonacci numbers,
 * please do so and let me know, it would let me sleep at night.
 */
#include <iostream>
#include <unordered_map>
#include <cassert>

long fibonacci(int index, std::unordered_map<int, long> &fibTable){
	// https://stackoverflow.com/a/6897865/11801882
	auto iter = fibTable.find(index);
	if(iter != fibTable.end()){
		return iter->second;
	}
	else {
		long a = fibonacci(index-1, fibTable);
		long b = fibonacci(index-2, fibTable);
		long answer = a + b;
		fibTable.insert({index, answer});
		return answer;
	}
}

long fibonacci(int index){
	std::unordered_map<int, long> fibTable;
	fibTable.insert({0,0});
	fibTable.insert({1,1});
	return fibonacci(index, fibTable);	
}


int main(int argc, char**argv){
	int index = 0x2f;
	long expected = 2971215073;
	long calculated = fibonacci(index);

	// https://stackoverflow.com/a/3692961/11801882
	assert(calculated == expected && "you are WRONG!!");

	std::cout << "Good job! Fibonacci index " << index << " is " << calculated << std::endl;
	return 0;	
}

File: dp/fib.py
#! /usr/bin/env python
""" I was going to implement this in `C++`, 
    but this algorithm is on the front page of Python website. (really!)
    So, much like my Haskell quicksort implementation,
    I feel obliged to include this bottom-up fibonacci memoization example."""

def fibonacci(n: int) -> int:
	a = 0
	b = 1
	for _ in range(n-1):
		a,b = b,a+b
	return b
	

if __name__ == "__main__":
	index = 0x2f
	expected = 2971215073
	calculated = fibonacci(index)
		
	assert expected == calculated, "YOU ARE WRONG!"
	print(f"Hurray! Fibonacci index {index} is {calculated}!")

